Basic federated support on Zephyr

erlingrj opened the pull request:
This PR mainly adds an example of federated execution on Zephyr using TcpIpChannel

erlingrj reviewed (commented):
Self-review done

tanneberger commented on examples/zephyr/basic_federated/federated_receiver/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+project(reactor-uc-zephyr)
+
+# TODO: Improve the way of pickng which Network channels to add to the build.
+set(NETWORK_POSIX_TCP ON CACHE BOOL "blah")
```
... "blah"

tanneberger commented on include/reactor-uc/macros.h:
```diff
@@ -237,16 +237,61 @@ typedef struct DelayedConnection DelayedConnection;
+#define DEFINE_FEDERATED_OUTPUT_CONNECTION(ConnectionName, BufferType)                                                 \
+  typedef struct {                                                                                                     \
+    FederatedOutputConnection super;                                                                                   \
+    BufferType buffer[1];                                                                                              \
```
does this suffer from the same problem that also action suffered, where values would be overwritten?

erlingrj replied:
No, it does not have a TriggerDataQueue, it is only a temporary place to store the value in case it is overwritten with other calls to `lf_set`

tanneberger commented on src/platform/posix/tcp_ip_channel.c:
```diff
@@ -17,6 +17,10 @@
 
 #include "proto/message.pb.h"
 
+#ifdef MIN
```
why this?

erlingrj replied:
Because sometimes `MIN` is already defined, it is the case for Zephyr, so then we undef to avoid errors

tanneberger commented on src/platform/posix/tcp_ip_channel.c:
```diff
@@ -196,11 +207,27 @@ void *TcpIpChannel_receive_thread(void *untyped_self) {
+  memset(&self->receive_thread_stack, 0, TCP_IP_CHANNEL_RECV_THREAD_STACK_SIZE);
+  if (pthread_attr_init(&self->receive_thread_attr) < 0) {
+    LF_ERR(NET, "pthread_attr_init failed with %d", errno);
+    throw("");
```
add error message

erlingrj replied:
Done!

erlingrj commented:
@tanneberger I have addressed your comments. CI is failing, but the fixes are in the `event-allocater` branch.  I am merging this now and planning on fixing everything with the event-allocator branch
